build: refactor icu-generic.gyp - kill path voodoo

refack opened the pull request:
1. Intention was to get `PRODUCT_DIR` so no need to do path voodoo

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
build, intl

mhdawson commented:
@srl295 as change is in ICU gyp file

bnoordhuis commented:
I think you can take this one step further and merge them with the other icutrim and genccode targets, they look functionally identical now.

fhinkel commented:
@refack Will you make the simplification that @bnoordhuis suggested or should we merge as is?

refack commented:
@fhinkel Give me a day to try and grok the rules.

refack commented:
@fhinkel, I broke... 😞
I refactored the conditions a bit, but I can't find a way to merge them...
I did find that they were actually broken because of excessive escaping so I added `'msvs_quote_cmd': 0`

refack commented:
You are all evil people, leaving me alone with this `gyp`.
It just sitting there... 😝 Mocking me... 🤖 "I'm so full of duplication",😈 "I'm so fragile I will take down your entire build"... 🔨

But I beat it. Well, just a bit.
Have fun reviewing...

refack commented:
[![Build status](https://ci.appveyor.com/api/projects/status/1rxtiumx5346aaas/branch/patch-2?svg=true)](https://ci.appveyor.com/project/refack/node/branch/patch-2) finaly green on windows

refack commented:
@nodejs/collaborators can I get a CI job on this?

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7132/

refack commented:
@nodejs/collaborators I think I cracked the last problem. Need a CI please...

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7156/

refack commented:
Ooof...
@Trott if you're there? Another one, please...

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7157/

refack commented:
🤦 🤦 🤦
ugh

Trott commented:
We should prioritize getting you onboarded to speed your work on this. I'm out and about right now but will send you an email later. Or send me one letting me know a date/time that would work well for you.

refack commented:
@jasnell I see you're cleaning house, could you CI this?

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/7207/

jasnell commented:
Lots of red in CI on this one

refack commented:
I've seen... I hate `GYP`... so frustrating...
I'm spinning off second part of the PR. Since the first part got the LGTM's and should be fairly stable, and it will unblock @indutny 's gyp.js efforts.

Need a new CI job just for df15e3b0277

gibfahn commented:
CI: https://ci.nodejs.org/job/node-test-commit/8893/

refack commented:
It's green. Quick land this, before `GYP` changes it's mind.

gibfahn commented:
@refack am I right in thinking that this change is what @srl295, @jasnell and @indutny approved, and that the changes you added later are now in https://github.com/nodejs/node/pull/12218?

If so then as long as @bnoordhuis doesn't object we can get this landed.

refack commented:
@gibfahn yes. What's left is what was here 10 days ago. The hard changes went to #12218.
This one seems like a minor change but it blocks `gyp.js` indutny/gyp.js#34

gibfahn commented:
One more Ci for luck and I'll land: https://ci.nodejs.org/job/node-test-commit/8920/

gibfahn commented:
Landed in https://github.com/nodejs/node/commit/e139dae157cca968e89de51ad435ded23f2b8be1

refack commented:
I was looking at this, reading the comments top to bottom, and was overjoyed again to see

> Landed in e139dae

MylesBorins commented:
This lands cleanly. Should it be cherry-picked to v6.x?

refack commented:
It's almost cosmetic, mainly to solve indutny/gyp.js#34.
If there's any risk, skip it.
